### ISSUE TITLE
Webapp progress

### ISSUE THREAD

**Issue by alejandrogallo** [papis/papis#292]
I might be working on a webapp for papis
in the `webapp` branch to use more ergonomically
papis on my ipad.

I wanted to ask who would be interested and if we
should brainstorm here some ideas.

Teaser:
![image](https://user-images.githubusercontent.com/12715358/99875468-39650800-2bf0-11eb-9dca-139f3be44fa5.png)

**Comment by alejandrogallo** [comment]
I abandoned this project in favour of a non-javascript based and simple solution for this, I'll be uploading to master this week a usable minimal implementation

**Comment by alejandrogallo** [comment]
The `papis-serve` command has been added, it is fairly usable already together with being extremely simple.
I'll be updating it in the near future and documenting it, whoever finds it useful is very welcome to discuss here
and adapt its use.
If you have a server with your papers and connect via your tablet to browse and so on it is very useful.

Thinks to do are:
- [ ] allow editing the fields of the paper
- [ ] implement a simple paper adder so that it is possible to add papers from a url (for instance arxiv which always works)
      with no javascript
- [ ] Implement the `explore` tab to have a frontend to crossref, arxiv and the backends exposed in `papis-explore`

**Comment by jghauser** [comment]
This does sound very interesting! I just tried it, but didn't get far:
- I just ran `papis serve` without specifying any options
- I tried to go (from the same computer) to the address shown: `https://localhost:8888`
- This gave me a `ERR_SSL_PROTOCOL_ERROR` in chromium or ` SSL_ERROR_RX_RECORD_TOO_LONG` in firefox (and a `code 400, message Bad request version` in the terminal)
- I tried with `http`, but that gave me the following error in chromium/firefox: `HTTP/1.0 400 Bad Request Server: BaseHTTP/0.6 Python/3.10.2 Date: Thu, 17 Feb 2022 21:38:19 GMT Content-Type: application/json {"message": "'list' object has no attribute 'split'"}`

Any idea what's going on? Do I need to somehow disable https for trying it out locally?

**Comment by alejandrogallo** [comment]
Interesing... in both my computers I don't have a problem, `papis serve` gets me the following:
```
INFO:papis:server:starting server in address https://localhost:8888
INFO:papis:server:press Ctrl-C to exit
INFO:papis:server:THIS COMMAND IS EXPERIMENTAL, expect bugs, feedback appreciated
INFO:papis:server:"GET / HTTP/1.1" 301 -
INFO:papis:server:"GET /library HTTP/1.1" 200 -
INFO:papis:server:"GET /favicon.ico HTTP/1.1" 400 -
```
do you get this `THIS COMMAND IS EXPERIMENTAL` part too?
I am not sure why you're getting `application/json` stuff there.

**Comment by alejandrogallo** [comment]
and it should look like this:

![image](https://user-images.githubusercontent.com/12715358/154582837-92fbea4e-d185-43b3-b1a6-4882366dbe21.png)

**Comment by jghauser** [comment]
It seems to start up normally, but then gets this `code 400` error. Here's the full output in the terminal. I'm gonna try troubleshoot a bit tomorrow!
```
❯ papis serve
INFO:papis:server:starting server in address https://localhost:8888
INFO:papis:server:press Ctrl-C to exit
INFO:papis:server:THIS COMMAND IS EXPERIMENTAL, expect bugs, feedback appreciated
INFO:papis:server:code 400, message Bad request version ('localhost\x00\x17\x00\x00ÿ\x01\x00\x01\x00\x00')
INFO:papis:server:"üÔFÎ¼Ã!'6@<*/§ j\QM9á$Ñ?nåGî×C
                                                 ñar3é¬²\Ðç± ZZÀ+À/À,À0Ì©Ì¨ÀÀ
```
If I try reloading the page, the code stays the same, but some of the details change:
```
NFO:papis:server:code 400, message Bad request version ('\x94§o')
INFO:papis:server:"üö|«È%èüÕ]xlX¼fÎÛ9*î [QÍ`
                                            §o" 400 -
INFO:papis:server:code 400, message Bad request version ('=ªÓ¨1%¦è')
ZëÄ%íE =ªÓ¨1%¦è" 400 -¡{åÔo[
```

Looks cryptic! :D

**Comment by alejandrogallo** [comment]
ok! I don't really understand what is going on there. I hope we figure it out and it becomes more usable.
Right now I just want it to be as simple as possible and if possible without using any javascript for the use case.
The upside is that you can browse it quite well with elinks... ;p

![image](https://user-images.githubusercontent.com/12715358/154589214-5f1bfff9-a577-4e0f-a111-70fbef87fafb.png)

**Comment by jghauser** [comment]
Oh totally missed your remark about `python3 -m http.server 8888`! This works without issues. I do see that this command creates an `http` link whereas with `papis serve` i always get an `https`, but i'm not sure that's got anything to do with the error.`papis serve --host 127.0.0.1` gives me the same error as just `papis serve`.

**Comment by wavefrontshaping** [comment]
Hi!

Just wanted to try the webapp, I have the exact same errors as @jghauser (Ubuntu 21.10)

`SSL_ERROR_RX_RECORD_TOO_LONG` on Firefox

and in the prompt:

```bash
INFO:papis:server:starting server in address https://localhost:8888
INFO:papis:server:press Ctrl-C to exit
INFO:papis:server:code 400, message Bad HTTP/0.9 request type ('\x16\x03\x01\x02\x00\x01\x00\x01ü\x03\x03')
INFO:papis:server:"ü
Ô±ÎÈ}K)}ÜG" 400 -
INFO:papis:server:"GET / HTTP/1.1" 200 -
INFO:papis:server:"GET / HTTP/1.1" 400 -
INFO:papis:server:"GET /favicon.ico HTTP/1.1" 400 -
INFO:papis:server:code 400, message Bad request versionG9H" 400 -aÐhRy\x8d\x8cG9H\x87')
INFO:papis:server:code 400, message Bad request version ('A\x00$\x13\x01\x13\x03\x13\x02À+À/Ì©Ì¨À,À0À')
INFO:papis:server:"üR=.âò¶/?DÌ¶w¥»^©\ÿ®9GÕpPÿ% ¬éùë¡ÐSº3é`À¬U±°^Þ`5¥4ÍA$À+À/Ì©Ì¨À,À0À" 400 -
INFO:papis:server:"GET / HTTP/1.1" 200 -
INFO:papis:server:"GET / HTTP/1.1" 400 -
INFO:papis:server:"GET /favicon.ico HTTP/1.1" 400 -
INFO:papis:server:code 400, message Bad request version ('localhost\x00\x17\x00\x00ÿ\x01\x00\x01\x00\x00')
INFO:papis:server:"üb1c(³Ëß¾M
}»Þ EÆ¶ó±7ÀFqÈPFú×?O ÊÊÀ+À/À,À0Ì©Ì¨ÀÀ
                                     	localhostÿ" 400 -
INFO:papis:server:code 400, message Bad request version ('localhost\x00\x17\x00\x00ÿ\x01\x00\x01\x00\x00')
INFO:papis:server:"üþ.8û2?[Y¨Pâ£N?³NIutALÚ {õÊ÷U¨5Eô>,
                                                      èR4sÏÝºæqmEä¨ **À+À/À,À0Ì©Ì¨ÀÀ
                                                                                    	localhostÿ" 400 -
INFO:papis:server:code 400, message Bad request syntax ('\x16\x03\x01\x02\x00\x01\x00\x01ü\x03\x03\x86eç\x96Eä5¥')
INFO:papis:server:"üeçEä5¥" 400 -
```
Is there something with the encoding/locales?

By the way, when I first tried to install from this branch, I had the following error:
```bash
error in papis setup command: Duplicate element EntryPoint(name='exec', value='papis.commands.exec:cli', group='papis.command') encountered.
```
as the line `"exec=papis.commands.exec:cli",` was duplicated in the setup.py

**Comment by alejandrogallo** [comment]
@wavefrontshaping thanks for testing, I don't understand what happens
but since @jghauser  said that his `python -m http.server` works then probably I have yet to setup something
correctly in the server, I'll take a look at the main function of the `http.server` implementation and then
get back to you later today.

By the way @wavefrontshaping, you should checkout the master branch,
I have the feeling that you are at the webapp branch, that branch is the old webapp
that I implemented in Javascript but it was too much of a hassle, so I did everything in the `serve` command
and shamelessly pushed it to master :|

**Comment by wavefrontshaping** [comment]
> but since @jghauser said that his python -m http.server works then probably I have yet to setup something

Same here, `python -m http.server` works without any error.

> I have the feeling that you are at the webapp branch, that branch is the old webapp

Indeed! Switched to main branch (same errors).

**Comment by alejandrogallo** [comment]
If you guys have time, could you change in `commands/serve.py`

from this (Line number 359)
```python
def _header_html(self) -> None:
    self.send_header("Content-Type", "text/html")
```

to this?


```python
def _header_html(self) -> None:
    self.send_header("Content-Type", "text/html;charset=utf-8")
```

I have the feeling the charset in the header is essential in both your systems,
that'd make sense.
I can also make the change and push, but maybe you can make the quick change

**Comment by wavefrontshaping** [comment]
Still the same here

**Comment by alejandrogallo** [comment]
oh my... can you post the result of `locale` on your machine?
For me is
```
locale
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC=de_AT.UTF-8
LC_TIME=de_AT.UTF-8
LC_COLLATE="en_US.UTF-8"
LC_MONETARY=de_AT.UTF-8
LC_MESSAGES="en_US.UTF-8"
LC_PAPER=de_AT.UTF-8
LC_NAME=de_AT.UTF-8
LC_ADDRESS=de_AT.UTF-8
LC_TELEPHONE=de_AT.UTF-8
LC_MEASUREMENT=de_AT.UTF-8
LC_IDENTIFICATION=de_AT.UTF-8
LC_ALL=
```

**Comment by wavefrontshaping** [comment]
```
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC=fr_FR.UTF-8
LC_TIME=fr_FR.UTF-8
LC_COLLATE="en_US.UTF-8"
LC_MONETARY=fr_FR.UTF-8
LC_MESSAGES="en_US.UTF-8"
LC_PAPER=fr_FR.UTF-8
LC_NAME=fr_FR.UTF-8
LC_ADDRESS=fr_FR.UTF-8
LC_TELEPHONE=fr_FR.UTF-8
LC_MEASUREMENT=fr_FR.UTF-8
LC_IDENTIFICATION=fr_FR.UTF-8
LC_ALL=
```

**Comment by alejandrogallo** [comment]
This is related
https://github.com/samuelclay/NewsBlur/issues/642

Just to be clear, you type `http://localhost:8888` or `https://localhost:8888`.
It should be `http://localhost:8888`, I realised the output said `https` now and that's
is really messy from my part, it should be http

**Comment by wavefrontshaping** [comment]
I tried them both, but I got a different error with `http`:

```
HTTP/1.0 400 Bad Request Server: BaseHTTP/0.6 Python/3.7.2 Date: Sat, 19 Feb 2022 08:23:58 GMT Content-Type: application/json {"message": "'list' object has no attribute 'split'"}
```

**Comment by jghauser** [comment]
Yeah, I've tried both as well, and I get the same errors as @wavefrontshaping.

**Comment by alejandrogallo** [comment]
question, your tags in your documents are a string or a list ? I think I know where the problem is,
probably some tags are a list for you guys so I'll handle the case accordingly and check if tags are lists
or a string etc.
Right now I have supposed (stupidly) that tags is a string of space-separated words

**Comment by alejandrogallo** [comment]
I have pushed a commit to that end, where I just check if `tags` is a string or a list

**Comment by wavefrontshaping** [comment]
It works!

**Comment by jghauser** [comment]
Yes, that solved it for me too! This is great, thanks @alejandrogallo !:tada:

Two things I feel could make the presentation a little cleaner:
- Maybe force some separation between the tags on the right and the title/author on the left.
- Force the size of the `doi`, `url` etc. buttons to some absolute value. 

Here's a screenshot to illustrate what I mean:
![20220219_12h09m56s_grim](https://user-images.githubusercontent.com/10319377/154798449-c5351638-1c8c-4c1d-b107-ff0b252847cd.jpeg)

**Comment by wavefrontshaping** [comment]
By the way, if you want to make it dynamic without too much trouble, maybe  [datatable.js](https://www.datatables.net/) can help you.
I did that some time ago but I remember it was not too difficult to manage with Python, you just create your table with some id and include a javascript function in the HTML template to interact with the table and handle the research and the pages.
(I did it for the [publication page](https://www.wavefrontshaping.net/publications/) on my website)

**Comment by alejandrogallo** [comment]
@jghauser 
Yes you're right, the format has to be worked on, specially for people with so many tags as yourself.
The other thing is that probably I'll make these links like to harvard ads and so on optional through configuration.
In principle I want to give the user as much configurabitliy as possible. For physics for instance ads is great, but I don't know
in other fields how good the harvard database is. The same with arxiv and so on.

@wavefrontshaping I originally did something like that in the old database but homebrew.
I'm not agains the idea. I would kind of like something similar to the current interface of papis,
i.e., you do a query to the database, you get X documents and then you can filter further.
No since having the rough webapp, I have several queries open in the browser statically and it would be somewhat also
useful to be able to fuzzy filter these static documents on the html. I have to see if `database.js` allows for that in a format
other than the one presented in their webpage (or yours), which also looks nice but I prefer multiline entries with
amenities such as tags and various urls not on a single line. Maybe it's doable to have then both at the same time

**Comment by wavefrontshaping** [comment]
@alejandrogallo I understand that a table object is not what you prefer, but I did a quick and dirty modification to make it a table and add `datatable.js`.
I simply created two columns and put what you have on the left-hand side in the first one and the rest in the second one. 

![image](https://user-images.githubusercontent.com/22410786/154802890-790cb522-13eb-457f-be37-aa61ce70b84b.png)

 It looks quite similar (and the columns have a constant width).


Look at my fork [wavefrontshaping/papis](https://github.com/wavefrontshaping/papis), by changing a few lines of codes I can have dynamic search!

**Comment by wavefrontshaping** [comment]
By the way, there is still a problem with the tags and the coma:

![image](https://user-images.githubusercontent.com/22410786/154804055-85aaf58f-8d9b-43a9-91c2-4cc6e6c4793c.png)

For instance *scattering media* should be one keyword and the comas should not appear.

The corresponding `.yml` file reads `tags: photoacoustic, focusing, scattering media`

**Comment by alejandrogallo** [comment]
@wavefrontshaping that is awesome, I'll take a look at this this week, by the way I updated the tags with a splitting
according to a configured regex, now it should work with space separated, comma separated or list tags

**Comment by wavefrontshaping** [comment]
Looks nice! Thanks! 

Just two details:
* about the tags, I sometimes put two-word tags, but it is rendered as two keywords, like 'scattering media':
![image](https://user-images.githubusercontent.com/22410786/155496274-ffde75c2-90c3-4c64-8f25-88fa8c8e6996.png)
I use coma separated string in the `.yml`, so it is here `scattering media, imaging`
Is there a way to have multiple-word tags recognized as one tag without breaking everything?

* When I click on the ref (here **Malavalli2019Scattering**, but does the same for all), I have an error page:
`HTTP/1.0 400 Bad Request Server: BaseHTTP/0.6 Python/3.7.2 Date: Thu, 24 Feb 2022 09:27:51 GMT Content-Type: application/json {"message": "('No document found with ref %s', 'ref:wang2018non-invasive')"}`
What is it supposed to do?

**Comment by jghauser** [comment]
@wavefrontshaping: My tags are lists which allows me to do multi word tags. But that would of course require you to rewrite all your files. I'm not sure what's the best way of dealing with all the different ways that people might use to organise their tags.

![20220224_10h35m49s_grim](https://user-images.githubusercontent.com/10319377/155498215-9bbd221e-9efe-4eb0-b416-51cae108d5a6.jpeg)

**Comment by wavefrontshaping** [comment]
Thanks for the comment, I guess it would be my last resort.

By the way, the tag search in the webapp does not seem to work for me, for instance with the url `http://localhost:8888/library/papers/query?q=tags:scattering`, I still see all the papers, not just the one with the tag `scattering`.

**Comment by alejandrogallo** [comment]
Hi, I wanted to ask if the webapp is basically usable for you two, for me I'm enjoying it, I would like more
functionality and so on but this can be done at some point in the future when I or someone has more time.
If you deem it usable I will prepare v0.12 of the papis release, I haven't released for a long time and I think
now it would be a good oportunity.

**Comment by wavefrontshaping** [comment]
Hi,
Nice work!
I still have the three issues I detailed in my previous posts:
* multiple word tags are rendered as multiple tags (using comma-separated string),
* when I click on a tag, It still displays all the articles, so the list is not filtered,
* I have an error page when I click on the reference name of the article (I guess I am supposed to see the details of the article there).

**Comment by jghauser** [comment]
Hi! I just spend a short time testing the webapp, here are some things i noticed.

- clicking on 'tags' and then a tag gives all entries rather than those tagged (although clicking on a tag from the list in 'all' works alright)
- when clicking a tag (from e.g. the 'all' page), it would be nice if the result page indicated the query (i.e. the tag in question)
- the citation ref label is not visually distinguishable from tags, making it seem like the date is randomly placed in the middle of a list of tags
- clicking on explore doesn't seem to do anything
- clicking on the citation ref label brings up the properties of document. this isn't obvious to see. i think maybe clicking anywhere on the entry (that doesn't do something else) should bring this up
- it would be nice to be able to adjust the width of the columns in the table

Thanks again for all the work! I'd be happy to do any more testing! I'm looking forward for the release, I'm planning on using it locally on my linux tablet :)

**Comment by wavefrontshaping** [comment]
I try to solve the issues I have on my fork, I am stuck on a weird problem:

I am working on the page to edit document yml file.
At first, when I clicked on the reference, I had an error page, it turned out that the GET page had a `("^/library/?([^/]+)?/document/(ref:.*)$"` instead of `("^/library/?([^/]+)?/document/ref:(.*)$"` (like for the POST), so the reference searched was : "ref:einstein1905" instead of "einstein1905".

That being solved, I have an error when submitting the changes.

In `update_page_document()`, it uses `db.query_dict(dict(ref=ref))` to find the document, but it returns nothing.
If I test manually, 
```python
db = papis.database.get('papers')
docs = db.get_all_documents()
db.query_dict(dict(ref='einstein1905'))
```
same for
```python
papis.database.cache.filter_documents(docs, 'ref: einstein1905')
```

On the other hand, in `page_document()` it uses `docs = papis.api.get_documents_in_lib(libname, ref)` and it works.
And manually, if I use `papis.api.get_documents_in_lib('papers', 'einstein1905')`, it does work.

What goes wrong here?

**Comment by alejandrogallo** [comment]
I don't understand what would be the problem,
so to sum up,
this does not work for you

```python
import papis.database
db = papis.database.get('papers')
docs = db.query_dict({"ref": "EinsteinCoeffihilbor1982"})
assert length(docs) == 1
docs = db.query("ref:EinsteinCoeffihilbor1982")
assert length(docs) == 1
```
whenever you have the following document

```
author: Hilborn, Robert C.
author_list:
- affiliation: []
  family: Hilborn
  given: Robert C.
doi: 10.1119/1.12937
files:
- output.pdf
issue: '11'
journal: American Journal of Physics
language: en
month: 11
pages: 982--986
publisher: American Association of Physics Teachers (AAPT)
ref: EinsteinCoeffihilbor1982
time-added: 2020-09-24-12:43:53
title: Einstein coefficients, cross sections, f values, dipole moments, and all that
type: article
url: http://dx.doi.org/10.1119/1.12937
volume: '50'
year: 1982
```

Do you have `ref:` well set? For me it works well, when 
i run the above code I get this document without issue.
I suspect your `ref` keyword is not well-put or maybe the python version?

The reason this would work
```python
papis.api.get_documents_in_lib('papers', 'einstein1905')
```
is because `einstein1905` matches everything in the `match-format` setting, but not specifically
only the reference.

For me, both work, I'd like to know then

- python version,
- `match-format` (`papis config match-format`)
- maybe a sample `info.yaml` file to see your `ref:` field

I hope we can figure this out!

Cheers

**Comment by wavefrontshaping** [comment]
* > maybe a sample info.yaml file to see your ref: field

With this example:
```yml
author: I. M. Vellekoop and A. P. Mosk
author_list:
- given_name: I.M.
  surname: Vellekoop
- given_name: A.P.
  surname: Mosk
doc_url:
- https://www.osapublishing.org/viewmedia.cfm?URI=ol-32-16-2309&seq=0
- http://doi.org/10.1364/OL.32.002309
doi: 10.1364/ol.32.002309
files:
- Vellekoop2007Focusing.pdf
first_name: Vellekoop
first_word: Focusing
journal: Optics Letters
last_name: Mosk
month: aug
number: '16'
pages: '2309'
publisher: The Optical Society
ref: vellekoop2007focusing
title: Focusing coherent light through opaque strongly scattering media
type: article
url: https://doi.org/10.1364%2Fol.32.002309
volume: '32'
year: '2007'
```

```python
import papis.database
db = papis.database.get('papers')
docs = db.query_dict({"ref": "vellekoop2007focusing"})
```
Returns nothing.



* > python version

Python 3.7.2

* > `match-format`

`{doc["journal"]}}{{doc["tags"]}}{{doc.subfolder}}{{doc["title"]}}{{doc["author"]}}{{doc["year"]}}`
(I use jinja)

**Comment by alejandrogallo** [comment]
I guess I imagine now what the problem is, you're using whoosh and
`"ref"` is not one of the schema fields, can it be?

If I put this in my config

```yaml
[papers]~                                                                       
whoosh-schema-fields = ["doi", "journal"]~                               
```

it does not work, if I put this

```yaml
[papers]~                                                                       
whoosh-schema-fields = ["doi", "ref", "journal"]~                               
```

it works. I don't use whoosh, I guess it would be a good idea
to add `ref` to the default `whoosh-schema-fields` anyways.

https://papis.readthedocs.io/en/latest/configuration.html#config-settings-whoosh-schema-fields

And I guess we might be in need of some further documentation about this.

Of course assuming this is the problem you're experiencing.

**Comment by wavefrontshaping** [comment]
Hi,
I do not use whoosh (it was not even installed), so I guess it came from somewhere else.
However, I do use jinja templating, can this be an issue?
I installed whoosh and added the line you gave, and now it works!

**Comment by wavefrontshaping** [comment]
ok, but now`papis.api.get_documents_in_lib(lib_name, ref)` does not work anymore, it seems that `get_documents_in_lib` does not look anymore in the refs...

**Comment by wavefrontshaping** [comment]
And with whoosh set, the tag search now works for me (before, clicking on one tag still displayed all the papers)
